main/util/src/mill/util/CoursierSupport.scala

```diff
@@ -208,60 +227,126 @@ trait CoursierSupport {
     }
   }
 
-  def resolveDependenciesMetadata(
+  @deprecated(
+    "Compatibility shim. Use the overload with parameter sameDependencyVersions instead",
+    "Mill after 0.11.7"
+  )
+  def resolveDependencies(
       repositories: Seq[Repository],
       deps: IterableOnce[coursier.Dependency],
       force: IterableOnce[coursier.Dependency],
+      sources: Boolean,
+      mapDependencies: Option[Dependency => Dependency],
+      customizer: Option[coursier.core.Resolution => coursier.core.Resolution],
+      ctx: Option[mill.api.Ctx.Log],
+      coursierCacheCustomizer: Option[
+        coursier.cache.FileCache[Task] => coursier.cache.FileCache[Task]
+      ],
+      resolveFilter: os.Path => Boolean
+  ): Result[Agg[PathRef]] =
+    resolveDependencies(
+      repositories = repositories,
+      deps = deps,
+      force = force,
+      sources = sources,
+      mapDependencies = mapDependencies,
+      customizer = customizer,
+      ctx = ctx,
+      coursierCacheCustomizer = coursierCacheCustomizer,
+      resolveFilter = resolveFilter,
+      sameDependencyVersions = Seq()
+    )
+
+  def resolveDependenciesMetadata(
+      repositories: Seq[Repository],
+      deps: IterableOnce[Dependency],
+      force: IterableOnce[Dependency],
       mapDependencies: Option[Dependency => Dependency] = None,
-      customizer: Option[coursier.core.Resolution => coursier.core.Resolution] = None,
+      customizer: Option[Resolution => Resolution] = None,
       ctx: Option[mill.api.Ctx.Log] = None,
       coursierCacheCustomizer: Option[
         coursier.cache.FileCache[Task] => coursier.cache.FileCache[Task]
-      ] = None
+      ] = None,
+      sameDependencyVersions: Seq[Set[(String, String)]] = Seq()
   ): (Seq[Dependency], Resolution) = {
 
-    val cachePolicies = coursier.cache.CacheDefaults.cachePolicies
+    val rootDeps = deps.iterator
+      .map(d => mapDependencies.fold(d)(_.apply(d)))
+      .toSeq
 
-    val forceVersions = force.iterator
-      .map(mapDependencies.getOrElse(identity[Dependency](_)))
+    val forceVersions: Map[Module, String] = force.iterator
+      .map(d => mapDependencies.fold(d)(_.apply(d)))
       .map { d => d.module -> d.version }
       .toMap
 
-    val start0 = Resolution()
-      .withRootDependencies(
-        deps.iterator.map(mapDependencies.getOrElse(identity[Dependency](_))).toSeq
-      )
-      .withForceVersions(forceVersions)
-      .withMapDependencies(mapDependencies)
```

Comment:
For this `withMapDependencies` I couldn't find an equivalent in the new API.

Reply:
Related?

* https://github.com/coursier/coursier/pull/3072

Reply:
I hadn't realized how deeply ingrained in Mill `mapDependencies` was (it's basically exposed to Mill users). On the [sbt side](https://github.com/coursier/sbt-coursier), it's only used internally, so these uses were converted to more explicit parameters in [`ResolutionParams`](https://github.com/coursier/coursier/blob/36341ddc7701bd38c3f8ac0e0e89ac116a2e2921/modules/coursier/shared/src/main/scala/coursier/params/ResolutionParams.scala) (`typelevel`, `doForceScalaVersion`, `doOverrideFullSuffix`, `selectedScalaVersion`).

https://github.com/coursier/coursier/pull/3072 adds back the possibility for API users to transform dependencies this way, yes.